File: vlib/regex/pcre/bench/bench.go
package main

import (
	"fmt"
	"regexp"
	"strings"
	"time"
)

const iterations = 10000
const warmup = 100

var shortText = "The quick brown fox jumps over the lazy dog 12345"
var mediumText = "The quick brown fox jumps over the lazy dog 12345. " +
	"Pack my box with five dozen liquor jugs. " +
	"How vexingly quick daft zebras jump! " +
	"The five boxing wizards jump quickly. " +
	"Sphinx of black quartz, judge my vow. " +
	"Two driven jocks help fax my big quiz."
var longText = strings.Repeat(mediumText, 20)
var emailText = "Contact us at user@example.com or admin@test.org for info. " +
	"Also try support@company.co.uk and test.user+tag@domain.com please."

type BenchResult struct {
	Name    string
	NsPerOp int64
	OpsSec  float64
}

func bench(name string, f func() int) BenchResult {
	for i := 0; i < warmup; i++ {
		f()
	}
	start := time.Now()
	for i := 0; i < iterations; i++ {
		f()
	}
	elapsed := time.Since(start)
	nsPerOp := elapsed.Nanoseconds() / int64(iterations)
	opsSec := float64(0)
	if nsPerOp > 0 {
		opsSec = 1e9 / float64(nsPerOp)
	}
	return BenchResult{Name: name, NsPerOp: nsPerOp, OpsSec: opsSec}
}

func main() {
	results := []BenchResult{}

	// 1. Literal match
	re1 := regexp.MustCompile(`quick`)
	results = append(results, bench("1. Literal match (short)", func() int {
		if re1.FindStringIndex(shortText) != nil {
			return 1
		}
		return 0
	}))

	// 2. Char class
	re2 := regexp.MustCompile(`[a-zA-Z]+`)
	results = append(results, bench("2. Char class [a-zA-Z]+", func() int {
		if re2.FindStringIndex(mediumText) != nil {
			return 1
		}
		return 0
	}))

	// 3. Alternation
	re3 := regexp.MustCompile(`fox|dog|cat|bird`)
	results = append(results, bench("3. Alternation (4 words)", func() int {
		if re3.FindStringIndex(shortText) != nil {
			return 1
		}
		return 0
	}))

	// 4. Find all digits
	re4 := regexp.MustCompile(`\d+`)
	results = append(results, bench("4. Find all digits", func() int {
		return len(re4.FindAllString(mediumText, -1))
	}))

	// 5. Groups
	re5 := regexp.MustCompile(`(\w+)\s+(\w+)`)
	results = append(results, bench("5. Groups (\\w+)\\s+(\\w+)", func() int {
		if re5.FindStringSubmatchIndex(mediumText) != nil {
			return 1
		}
		return 0
	}))

	// 6. Email
	re6 := regexp.MustCompile(`[\w.]+@[\w]+\.[\w]+`)
	results = append(results, bench("6. Email pattern", func() int {
		return len(re6.FindAllString(emailText, -1))
	}))

	// 7. Long text
	re7 := regexp.MustCompile(`quickly`)
	results = append(results, bench("7. Long text literal scan", func() int {
		return len(re7.FindAllString(longText, -1))
	}))

	// 8. Replace all
	re8 := regexp.MustCompile(`\d+`)
	results = append(results, bench("8. Replace all digits", func() int {
		return len(re8.ReplaceAllString(mediumText, "NUM"))
	}))

	// 9. Dot-star
	re9 := regexp.MustCompile(`.*fox.*dog`)
	results = append(results, bench("9. Dot-star greedy", func() int {
		if re9.MatchString(shortText) {
			return 1
		}
		return 0
	}))

	// 10. Compile speed
	results = append(results, bench("10. Compile simple", func() int {
		_, err := regexp.Compile(`\w+@\w+\.\w+`)
		if err != nil {
			return 0
		}
		return 1
	}))

	// 11. Lookahead (Go doesn't support — skip)
	results = append(results, BenchResult{Name: "11. Lookahead", NsPerOp: -1, OpsSec: 0})

	// 12. Backreference (Go doesn't support — skip)
	results = append(results, BenchResult{Name: "12. Backref", NsPerOp: -1, OpsSec: 0})

	// Print
	fmt.Println(strings.Repeat("=", 70))
	fmt.Println("                 GO REGEX BENCHMARK (regexp)")
	fmt.Println(strings.Repeat("=", 70))
	fmt.Printf("%-35s %12s %14s\n", "Benchmark", "ns/op", "ops/sec")
	fmt.Println(strings.Repeat("-", 70))
	for _, r := range results {
		if r.NsPerOp < 0 {
			fmt.Printf("%-35s %12s %14s\n", r.Name, "N/A", "unsupported")
		} else {
			fmt.Printf("%-35s %12d %14.0f\n", r.Name, r.NsPerOp, r.OpsSec)
		}
	}
	fmt.Println(strings.Repeat("=", 70))
}

File: vlib/regex/pcre/bench/bench.py
#!/usr/bin/env python3
import re
import time

ITERATIONS = 10000
WARMUP = 100

short_text = "The quick brown fox jumps over the lazy dog 12345"
medium_text = (
    "The quick brown fox jumps over the lazy dog 12345. "
    "Pack my box with five dozen liquor jugs. "
    "How vexingly quick daft zebras jump! "
    "The five boxing wizards jump quickly. "
    "Sphinx of black quartz, judge my vow. "
    "Two driven jocks help fax my big quiz."
)
long_text = medium_text * 20
email_text = (
    "Contact us at user@example.com or admin@test.org for info. "
    "Also try support@company.co.uk and test.user+tag@domain.com please."
)


def bench(name, func):
    for _ in range(WARMUP):
        func()
    start = time.perf_counter_ns()
    for _ in range(ITERATIONS):
        func()
    elapsed = time.perf_counter_ns() - start
    ns_per_op = elapsed // ITERATIONS
    ops_sec = 1e9 / ns_per_op if ns_per_op > 0 else 0
    return (name, ns_per_op, ops_sec)


results = []

# 1. Literal
p1 = re.compile(r"quick")
results.append(bench("1. Literal match (short)", lambda: p1.search(short_text)))

# 2. Char class
p2 = re.compile(r"[a-zA-Z]+")
results.append(bench("2. Char class [a-zA-Z]+", lambda: p2.search(medium_text)))

# 3. Alternation
p3 = re.compile(r"fox|dog|cat|bird")
results.append(bench("3. Alternation (4 words)", lambda: p3.search(short_text)))

# 4. Find all
p4 = re.compile(r"\d+")
results.append(bench("4. Find all digits", lambda: p4.findall(medium_text)))

# 5. Groups
p5 = re.compile(r"(\w+)\s+(\w+)")
results.append(bench("5. Groups (\\w+)\\s+(\\w+)", lambda: p5.search(medium_text)))

# 6. Email
p6 = re.compile(r"[\w.]+@[\w]+\.[\w]+")
results.append(bench("6. Email pattern", lambda: p6.findall(email_text)))

# 7. Long text
p7 = re.compile(r"quickly")
results.append(bench("7. Long text literal scan", lambda: p7.findall(long_text)))

# 8. Replace
p8 = re.compile(r"\d+")
results.append(bench("8. Replace all digits", lambda: p8.sub("NUM", medium_text)))

# 9. Dot-star
p9 = re.compile(r".*fox.*dog")
results.append(bench("9. Dot-star greedy", lambda: p9.match(short_text)))

# 10. Compile
results.append(bench("10. Compile simple", lambda: re.compile(r"\w+@\w+\.\w+")))

# 11. Lookahead
p11 = re.compile(r"\w+(?=\.)")
results.append(bench("11. Lookahead \\w+(?=\\.)", lambda: p11.search(email_text)))

# 12. Backreference
doubled = "the the quick quick fox fox jumps over the lazy dog"
p12 = re.compile(r"(\w+)\s+\1")
results.append(bench("12. Backref (\\w+)\\s+\\1", lambda: p12.findall(doubled)))

print("=" * 70)
print(f"{'':>15}PYTHON REGEX BENCHMARK (re module)")
print("=" * 70)
print(f"{'Benchmark':<35} {'ns/op':>12} {'ops/sec':>14}")
print("-" * 70)
for name, ns, ops in results:
    print(f"{name:<35} {ns:>12} {ops:>14.0f}")
print("=" * 70)
